File: qlib-service/deploy_pkg/decision_engine/data/fuyao.py
"""Same-source Fuyao market-data adapter for training archives."""

from __future__ import annotations

import concurrent.futures
import datetime as dt
import json
import math
import os
import urllib.parse
import urllib.request
from zoneinfo import ZoneInfo


BASE_URL = "https://fuyao.aicubes.cn"
PAGE_SIZE = 1000
MAX_PAGES = 10
SOURCE = "THS_FUYAO"
ZONE = ZoneInfo("Asia/Shanghai")


def configured(env=None):
    source = os.environ if env is None else env
    return bool(str(source.get("FUYAO_API_KEY") or "").strip())


def _number(value):
    if value in (None, "", "-") or isinstance(value, bool):
        return None
    try:
        number = float(value)
    except (TypeError, ValueError):
        return None
    return number if math.isfinite(number) else None


def _trade_date(timestamp):
    value = _number(timestamp)
    if value is None or value <= 0:
        return None
    return dt.datetime.fromtimestamp(
        value / 1000,
        tz=ZONE,
    ).strftime("%Y%m%d")


def request_json(path, query, *, env=None, timeout=12):
    source = os.environ if env is None else env
    key = str(source.get("FUYAO_API_KEY") or "").strip()
    if not key:
        return None
    url = (
        BASE_URL
        + path
        + "?"
        + urllib.parse.urlencode(query)
    )
    request = urllib.request.Request(
        url,
        headers={
            "Accept": "application/json",
            "X-api-key": key,
        },
    )
    with urllib.request.urlopen(request, timeout=timeout) as response:
        payload = json.loads(response.read().decode("utf-8"))
    if int(payload.get("code", -1)) != 0:
        raise RuntimeError(
            f"扶摇行情业务错误 {payload.get('code', 'UNKNOWN')}"
        )
    data = payload.get("data")
    if not isinstance(data, dict) or not isinstance(data.get("item"), list):
        raise ValueError("扶摇行情响应结构无效")
    return data


def fetch_snapshot_page(
    offset,
    *,
    page_size=PAGE_SIZE,
    env=None,
    request=request_json,
):
    return request(
        "/api/a-share/prices/snapshot",
        {
            "limit": int(page_size),
            "offset": int(offset),
        },
        env=env,
    )


def fetch_full_snapshot(
    *,
    env=None,
    workers=6,
    fetch_page=None,
):
    if not configured(env):
        return None
    load_page = fetch_page or (
        lambda offset: fetch_snapshot_page(offset, env=env)
    )
    first = load_page(0)
    total = int((first or {}).get("total") or 0)
    if total < 800:
        raise ValueError("扶摇全市场股票总数异常")
    page_count = math.ceil(total / PAGE_SIZE)
    if not 1 <= page_count <= MAX_PAGES:
        raise ValueError("扶摇全市场分页数异常")
    pages = {0: first}
    offsets = [
        index * PAGE_SIZE
        for index in range(1, page_count)
    ]
    with concurrent.futures.ThreadPoolExecutor(
        max_workers=max(1, min(int(workers), 12)),
    ) as executor:
        futures = {
            executor.submit(load_page, offset): offset
            for offset in offsets
        }
        for future in concurrent.futures.as_completed(futures):
            pages[futures[future]] = future.result()
    rows = []
    timestamps = []
    for offset in [index * PAGE_SIZE for index in range(page_count)]:
        page = pages[offset]
        items = page.get("item") if isinstance(page, dict) else None
        expected = min(PAGE_SIZE, total - offset)
        if not isinstance(items, list) or len(items) < expected:
            raise ValueError(
                f"扶摇全市场分页不完整: {offset}/{total}"
            )
        rows.extend(items)
        if _number(page.get("timestamp")) is not None:
            timestamps.append(page["timestamp"])
    unique = {
        str(row.get("ticker") or ""): row
        for row in rows
        if str(row.get("ticker") or "").isdigit()
        and len(str(row.get("ticker"))) == 6
    }
    if len(unique) != total:
        raise ValueError(
            f"扶摇全市场快照不完整: {len(unique)}/{total}"
        )
    target = _trade_date(max(timestamps)) if timestamps else None
    if not target:
        raise ValueError("扶摇全市场快照缺少交易日期")
    mapped = {}
    for code, row in unique.items():
        close = _number(row.get("last_price"))
        open_price = _number(row.get("open_price"))
        high = _number(row.get("high_price"))
        low = _number(row.get("low_price"))
        if (
            close is None
            or open_price is None
            or high is None
            or low is None
            or min(close, open_price, low) <= 0
            or high < max(open_price, close)
            or low > min(open_price, close)
        ):
            continue
        mapped[code] = {
            "date": target,
            "code": code,
            "open": open_price,
            "high": high,
            "low": low,
            "close": close,
            "preClose": _number(row.get("prev_price")),
            "volume": _number(row.get("volume")) or 0.0,
            "amount": _number(row.get("turnover")) or 0.0,
            "source": SOURCE,
        }
    if len(mapped) < 800:
        raise ValueError("扶摇全市场有效行情覆盖不足")
    return {
        "date": target,
        "total": total,
        "rows": mapped,
        "coverage": len(mapped) / total,
        "source": SOURCE,
    }

File: api/_fuyao.js
import { beijingDayKey } from '../shared/tradingCalendar.js'

export const FUYAO_BASE_URL = 'https://fuyao.aicubes.cn'
export const FUYAO_SOURCE = '同花顺扶摇'

function finite(value) {
  if (value == null || value === '') return null
  const number = Number(value)
  return Number.isFinite(number) ? number : null
}

function positive(value) {
  const number = finite(value)
  return number != null && number > 0 ? number : null
}

function apiKey(env) {
  return String(env?.FUYAO_API_KEY || '').trim()
}

export function fuyaoConfigured(env = process.env) {
  return apiKey(env).length > 0
}

export function toFuyaoThscode(code) {
  const normalized = String(code || '').trim()
  if (!/^\d{6}$/.test(normalized)) return null
  if (/^(4|8|92)/.test(normalized)) return `${normalized}.BJ`
  if (/^(5|6|9)/.test(normalized)) return `${normalized}.SH`
  return `${normalized}.SZ`
}

function endpoint(path, query = {}) {
  const url = new URL(path, FUYAO_BASE_URL)
  for (const [key, value] of Object.entries(query)) {
    if (value == null || value === '') continue
    url.searchParams.set(key, String(value))
  }
  return url
}

async function requestFuyao(
  path,
  query,
  {
    env = process.env,
    fetchImpl = globalThis.fetch,
    timeoutMs = 5000,
  } = {},
) {
  const key = apiKey(env)
  if (!key) return null
  const controller = new AbortController()
  const timer = setTimeout(
    () => controller.abort(),
    Math.max(1000, Number(timeoutMs) || 5000),
  )
  try {
    const response = await fetchImpl(endpoint(path, query), {
      headers: {
        Accept: 'application/json',
        'X-api-key': key,
      },
      signal: controller.signal,
    })
    if (!response.ok) {
      throw new Error(`扶摇行情HTTP ${response.status}`)
    }
    const payload = await response.json()
    if (Number(payload?.code) !== 0) {
      const code = Number.isFinite(Number(payload?.code))
        ? Number(payload.code)
        : 'UNKNOWN'
      throw new Error(`扶摇行情业务错误 ${code}`)
    }
    if (!payload?.data || !Array.isArray(payload.data.item)) {
      throw new Error('扶摇行情响应结构无效')
    }
    return payload.data
  } finally {
    clearTimeout(timer)
  }
}

export function mapFuyaoQuote(row = {}, timestamp = null) {
  const volume = positive(row.volume)
  const amount = positive(row.turnover)
  return {
    code: String(row.ticker || '').trim(),
    name: '',
    source: FUYAO_SOURCE,
    price: positive(row.last_price),
    pct: finite(row.price_change_ratio_pct),
    chg: finite(row.price_change),
    turnover: null,
    volRatio: null,
    mainInflow: null,
    retailInflow: null,
    main5dInflow: null,
    retail5dInflow: null,
    mainRatio: null,
    volume,
    amount,
    vwap: volume && amount ? amount / volume : null,
    high: positive(row.high_price),
    low: positive(row.low_price),
    open: positive(row.open_price),
    prevClose: positive(row.prev_price),
    tradeDate: finite(timestamp) != null
      ? beijingDayKey(Number(timestamp))
      : null,
    industry: null,
    thscode: String(row.thscode || '').trim() || null,
  }
}

export async function fetchFuyaoQuotes(
  codes,
  options = {},
) {
  const normalized = [...new Set(
    (Array.isArray(codes) ? codes : [])
      .map(toFuyaoThscode)
      .filter(Boolean),
  )]
  if (!normalized.length || !fuyaoConfigured(options.env)) return []
  const chunks = []
  for (let index = 0; index < normalized.length; index += 100) {
    chunks.push(normalized.slice(index, index + 100))
  }
  const responses = await Promise.all(chunks.map((chunk) =>
    requestFuyao(
      '/api/a-share/prices/snapshot',
      { thscodes: chunk.join(',') },
      options,
    )
  ))
  return responses.flatMap((data) =>
    (data?.item || []).map((row) =>
      mapFuyaoQuote(row, data.timestamp)
    )
  )
}

function historicalWindow(limit, now) {
  const end = Number(now) || Date.now()
  const days = Math.min(3650, Math.max(45, limit * 3))
  return {
    start: end - days * 24 * 60 * 60 * 1000,
    end,
  }
}

export async function fetchFuyaoDailyKline(
  code,
  klt = '101',
  limit = 120,
  options = {},
) {
  if (String(klt) !== '101' || !fuyaoConfigured(options.env)) {
    return null
  }
  const thscode = toFuyaoThscode(code)
  if (!thscode) return null
  const boundedLimit = Math.max(
    1,
    Math.min(500, Math.trunc(Number(limit) || 120)),
  )
  const window = historicalWindow(boundedLimit, options.now)
  const data = await requestFuyao(
    '/api/a-share/prices/historical',
    {
      thscode,
      interval: '1d',
      start: window.start,
      end: window.end,
      adjust: 'forward',
    },
    options,
  )
  const candles = (data?.item || [])
    .map((row) => ({
      date: finite(row.date_ms) != null
        ? beijingDayKey(Number(row.date_ms))
        : '',
      open: positive(row.open_price),
      close: positive(row.close_price),
      high: positive(row.high_price),
      low: positive(row.low_price),
      volume: positive(row.volume) || 0,
      amount: positive(row.turnover) || 0,
      pct: 0,
      turnover: null,
    }))
    .filter((row) =>
      row.date
      && row.open > 0
      && row.close > 0
      && row.high > 0
      && row.low > 0
    )
    .sort((left, right) => left.date.localeCompare(right.date))
    .slice(-boundedLimit)
  for (let index = 1; index < candles.length; index += 1) {
    const previous = candles[index - 1].close
    candles[index].pct = previous > 0
      ? +(((candles[index].close / previous) - 1) * 100).toFixed(2)
      : 0
  }
  return candles.length
    ? {
        name: '',
        candles,
        source: FUYAO_SOURCE,
      }
    : null
}
